### DISCUSSION TITLE
Valhalla graph tiles and dynamic tile loading

### DISCUSSION THREAD

**Discussion by NitinMehra0019** [valhalla/valhalla#4537] [Q&A] [answered]
Hi, I want to know the basic detail like what happen when we inject OSM data to the valhalla engine like data distributed into tiles how this distribution happens and also when a routing request happen then how the dynamic tile loading happens how these tiles are choosen to be loaded. 
Thankyou.

**Comment by ImreSamu** [comment]
The Valhalla documentation contains a lot of basic information about the structure of graph tiles.

 Have you had a chance to look through them?

* https://valhalla.github.io/valhalla/mjolnir/why_tiles/
* https://valhalla.github.io/valhalla/tiles/
   * Get the Level or Hierarchy
   * Get the Latitude and Longitude from an ID
   * Get the Tile ID from a Latitude, Longitude, and Level
   * Get Tile ID from an ID
* https://valhalla.github.io/valhalla/decoding/
* https://valhalla.github.io/valhalla/route_overview/
* ...

**Reply by NitinMehra0019** [reply]
Thanks @ImreSamu for your response I have already read the above docs but didn't get the actual flow how OSM data gets converted into graph tiles and when routing request happen how it is decided which tiles are going to be loaded.

**Reply by ImreSamu** [reply, marked as answer]
I only have an user-level knowledge of Valhalla, so I'm not able to help much with the details.

> how OSM data gets converted into graph tiles 

It's beneficial to understand the data model and tagging of OpenStreetMap. 
* https://wiki.openstreetmap.org/wiki/Highways
* https://wiki.openstreetmap.org/wiki/Key:highway

Most business rules are outsourced into Lua codes, which are easy to modify. 
* https://github.com/valhalla/valhalla/blob/master/lua/admin.lua
* https://github.com/valhalla/valhalla/blob/master/lua/graph.lua 

In the config check the "mjolnir" related parameters:
* https://github.com/valhalla/valhalla/blob/master/scripts/valhalla_build_config

The output tile proto format: https://github.com/valhalla/valhalla/tree/master/proto

in the program code/documentation search for the "mjolnir"
*  https://github.com/search?q=repo%3Avalhalla%2Fvalhalla+mjolnir&type=code
*  https://github.com/valhalla/valhalla/tree/master/src/mjolnir
*  https://github.com/valhalla/valhalla/tree/master/valhalla/mjolnir
 
If I needed to delve deeper, I would create a mini topology of about 6-8 streets (input) and try to better understand the algorithm based on the finished tiles (output). At least, this approach usually helps me.

Another resource that helped me understand this better was the PgRouting tutorials, which illustrate the essence of routing topology quite well. For instance, they explain what a node is, what an edge is, and how the routing algorithm works.

Since I'm not familiar with your background knowledge or the specific aim of your question, I'm afraid I can't provide much more assistance than this.

> when routing request happen how it is decided which tiles are going to be loaded.

As for when a routing request happens and how it decides which tiles to load, 
I assume that the request allows for the determination of the affected area (Bounding Box). 

There's an algorithm provided for calculating the tiles involved: 
["Get Tiles That Intersect a Bounding Box" (a Python algorithm).](https://valhalla.github.io/valhalla/tiles/#get-tiles-that-intersect-a-bounding-box) 
* `tiles_for_bounding_box(-74.251961,40.512764,-73.755405,40.903125)` ->  Returns a list of tiles for the NYC bounding box at each level.
